Empty Pull Request for Review

carschno opened the pull request.

jiqicn committed:
Update gitignore file; remove old unit tests.

jiqicn committed:
Add a placeholder test.

n400peanuts committed:
Merge pull request #17 from covid19ABM/refactor

Refactor

Commit:
removed old tutorial

Commit:
model class updated with tutorial

Commit:
tutorial updated

jiqicn committed:
Add function of reading features and corresponding value distributions from JSON file; make a sample feature file.

jiqicn committed:
Adapt Individual.populate to accept feature files as input.

jiqicn committed:
Add function to automatically create empty hypothesis files.

jiqicn committed:
Merge pull request #18 from covid19ABM/model

Model class updated

jiqicn committed:
Merge branch 'main' of https://github.com/covid19ABM/mhm into dev_hyp

jiqicn committed:
Merge from main.

Update the model class; reorganize the project; separate the gender column in example parameter files to gender_M and gender_F.

jiqicn committed:
Refactor the populate fucntion to indicate the directory of keeping parameter files.

jiqicn committed:
Add package methods for creating empty hypothesis files.

jiqicn committed:
Refactor functions for choosing and taking actions corresponding to the previous changes.

jiqicn committed:
Add size, steps, and lockdown intervals to the model parameter files.

jiqicn committed:
Update mhm/__init__.py

Co-authored-by: Eva Viviani <eva_viviani@msn.com>

jiqicn committed:
Update mhm/__init__.py

Co-authored-by: Eva Viviani <eva_viviani@msn.com>

Commit:
model now accepts the directory of the hypotheses folder among the arguments in the setup()

Commit:
tutorial updated to reflect changes

Commit:
parameters example files changed. Model class updated to reflect the changes in the individual class

Commit:
removed additional notebook

n400peanuts committed:
Merge pull request #22 from covid19ABM/dev_hyp

Merge branch dev_hyp of @jiqicn into main

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

Commit:
changed name of the package

n400peanuts committed:
Merge pull request #25 from covid19ABM/n400peanuts-patch-1

N400peanuts update readme

Commit:
added pyproject.toml

Commit:
added picture for the package

Commit:
readme updated

Commit:
updated branch with latest name of the package

Commit:
modified docs

Commit:
authors updated

Commit:
docs updated

Commit:
documentation yaml updated

Commit:
autoapi directory set to comma

n400peanuts committed:
Merge pull request #27 from covid19ABM/poetry_toml

Poetry toml

Commit:
conf docs updated

n400peanuts committed:
Merge pull request #28 from covid19ABM/poetry_toml

conf docs updated

Commit:
docs updated

Commit:
added img folder in docs

Commit:
readme updated

Commit:
added myst-parser

n400peanuts committed:
Merge pull request #29 from covid19ABM/docs_update

Docs update

Commit:
readme updated

Commit:
readme updated

Commit:
readme updated

Commit:
readme updated

Commit:
readme updated

Commit:
readme and pyproject.toml updated
/

Commit:
pyproject and readme updated

Commit:
readme updated

Commit:
Merge remote-tracking branch 'origin/main' into docs_

Commit:
corrected typo in the readme

Commit:
update readme.rst

Commit:
added source_suffix = [.rst, .md]

Commit:
Merge branch 'main' into docs_

n400peanuts committed:
Merge pull request #31 from covid19ABM/docs_

Docs

Commit:
error found in column order

Commit:
added test to validate csvs

Commit:
validate csvs does not test types

Commit:
validate_csvs.py now starts with "test" otherwise pytest does not find it..

Commit:
add correct test file

Commit:
install pandas in the yaml test file

jiqicn committed:
Merge pull request #32 from covid19ABM/test-develop

testing matrices

jiqicn committed:
Read the hypothesis files as desired.

By reading the hypothesis file, the result dataframe should have exactly the same number and order of columns as  and the same number and order of rows corresponding to the  variable, that is given in the  file.

jiqicn committed:
Extract the code of feature encoding as a separate function.

jiqicn committed:
Add the parameter files validation function.

jiqicn committed:
Add tests for the validation function.

jiqicn committed:
Add pytest.ini file.

jiqicn committed:
Remove value our-of-range validation.

n400peanuts committed:
Merge pull request #34 from covid19ABM/validation

Merge validation

n400peanuts committed:
Update README.md

there is no need to have poetry to install this package

Commit:
convert README.md in README.rst at every push

Commit:
pandoc installed

Commit:
anoter approach to install pandoc on ubuntu

Commit:
push newly created rst file back to the repo

actions-user committed:
Update README.rst

Commit:
added branch main in the workflow

Commit:
docs workflow now runs on main

n400peanuts committed:
Merge pull request #40 from covid19ABM/docs_update

Docs update

Commit:
doc about prob matrices

n400peanuts committed:
Merge pull request #42 from covid19ABM/add_documentation

doc about prob matrices

Commit:
example

Commit:
added nbsphinx to modules to be installed

n400peanuts committed:
Merge pull request #43 from covid19ABM/add_documentation

add notebook example in the documentation -- this is to clarify what lockdown matrices do

Commit:
added sphinx-gallery

Commit:
ok
Merge branch 'main' into add_documentation

n400peanuts committed:
Merge pull request #44 from covid19ABM/add_documentation

Add notebooks into documentation

Commit:
added gallery for notebooks

Commit:
removed pandoc madness

Commit:
errata corrige so pandoc is necessary apparently

jiqicn committed:
Fix #51.

Add sampling function from the IPF weights; add setup.py file so that the tutorial notebook will work properly.

Astrid-p committed:
Add files via upload

Hi Eva, 

Please find the updated matrices (17July) in the parameters folder. <3

jiqicn committed:
Add the unit test of the ipf sampling function.

Commit:
test ipf sampling against the crosstabs of the LA

Commit:
removed test_individual.py

Commit:
lockdown matrices updated

Commit:
update on matrices

Commit:
update matrices

n400peanuts committed:
Merge pull request #56 from covid19ABM/update_matrices_17july

Update matrices 17july

Commit:
update
Merge branch 'main' into ipf_sample

n400peanuts committed:
Update python-package.yml

Commit:
corrected return with assert

Commit:
ops corrected typo

Commit:
added the weights in the correct folder

Commit:
removed erroneuos print

Commit:
corrected selection of files

Commit:
removed setup.py (Ji I think put it back) and update toml with scipy + its version

Commit:
removed _keys as suggested by Carsten

Commit:
removed extra white space

Commit:
renamed crosstabs with more semantically transparent names

Commit:
implemented Carsten"s suggestion

Commit:
added intro about the tests in both files

Commit:
changed assert with raise ValueError as suggested by Carsten

Commit:
made assert checks more robust

Commit:
added matrix of actions effect on mental health

Commit:
update individual and model parameters

Commit:
added tqdm to toml

Commit:
yaml workflow updated

Commit:
solving dependencies version issues

Commit:
installed flake8 via toml

Commit:
run flake8 with poetry

Commit:
added more comments and specification of variables in choose_actions_on_lockdown

Commit:
added test for choosing actions based on lockdown and fixed few things in individual

Commit:
removed typo in the path

Commit:
added progress bar to run simulation

Commit:
added test for populate, and fixed typo in the matrices

Commit:
refined populate_ipf function in individual class and relative model functions

Commit:
added new test for ipf weights that checks the function populate of the individual class

Commit:
latest improves on the code

Commit:
tutorial updated

Commit:
removed simulation results csv

n400peanuts committed:
Merge pull request #60 from covid19ABM/refactor

Update comma to reflect the latest features selected by the LA

n400peanuts committed:
Added Untitled Diagram.drawio

n400peanuts committed:
Update comma diagram.drawio

n400peanuts committed:
Update Untitled Diagram.drawio

n400peanuts committed:
Rename Untitled Diagram.drawio to comma_diagram.drawio

n400peanuts committed:
Added Untitled Diagram.drawio.svg

n400peanuts committed:
Rename Untitled Diagram.drawio.svg to comma_diagram.drawio.svg

n400peanuts committed:
Update README.md

n400peanuts committed:
Update README.md

n400peanuts committed:
Delete comma_diagram.drawio

n400peanuts committed:
Merge pull request #61 from covid19ABM/draw.io

Draw.io

n400peanuts committed:
Update README.md

n400peanuts committed:
Merge pull request #62 from covid19ABM/n400peanuts-patch-1

Update README.md

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+
+class Individual:
+    _features = pd.DataFrame()
+    _status = pd.DataFrame()
```
I understand `_features` and `_status` are global and store the respective values for each initialized `Inidividual` object, correct?
I would find it clearer if each `Individual` object had their own `_features` and `_status` variable, initialized upon object instantiation in `__init__()`.

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+        Returns:
+            pd.Series: the current status of the agent
+        """
+        return self._status.loc[self.id]
```
Related to my comment above, these two methods could just return `self._features`/`self._status`.

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+        """
+        fpath_params_lockdown = os.path.join(
+            self.dir_params, 'lockdown_%s.csv' % lockdown)
+        params_lockdown = self._read_hypothesis(fpath_params_lockdown)
```
I am not clear yet when exactly `choose_actions_on_lockdown()` is called, but it (re-)reads the hypothesis from disk every time. Unless there is an issue with RAM, I would instead load the hypothesis once per individual upon instantiation in the `__init__()` method.

n400peanuts replied:
oh wow, you're right this is called at every step within the `choose_actions_on_lockdown` -- good catch!

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+            fpath_params_status = os.path.join(
+                self.dir_params, 'actions_effects_on_%s.csv' % s)
+            params_status = self._read_hypothesis(fpath_params_status)
+            result = params_status.dot(self.get_features()).dot(actions)
```
Shouldn't this use `self.chosen_action`? Why is `actions` passed into this function?

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+        for s in status:
+            fpath_params_status = os.path.join(
+                self.dir_params, 'actions_effects_on_%s.csv' % s)
+            params_status = self._read_hypothesis(fpath_params_status)
```
Re-reading the hypothesis from disk should be avoided (see comment above).

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+            'selfrated_health_poor',
+            'critical_job_no',
+            'critical_job_yes'
+        ]
```
Perhaps move `all_possible_cols` out of this function for readability. It could live in `../settings.py` and be imported here.

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+            'critical_job_yes'
+        ]
+
+        # hot-one encoding
```
```suggestion
        # one-hot encoding
```
:-)

carschno commented on comma/individual.py:
```diff
@@ -0,0 +1,272 @@
+                'selfrated_health_poor',
+                'critical_job_no',
+                'critical_job_yes'
+            ]
```
This looks like a duplication of `all_possible_cols` above; one more reason to move it to somewhere else, and or they will be out of sync at some point in the future.

carschno commented on comma/model.py:
```diff
@@ -0,0 +1,190 @@
+            self.agents = Individual.populate(size, dir_params)
+        return self.agents
+
+    def param_file_validation(self, dir_params: str):
```
Style remark: functions should be named using a verb to indicate an action they implement:
 ```suggestion
    def validate_param_file(self, dir_params: str):
```

carschno commented on comma/model.py:
```diff
@@ -0,0 +1,190 @@
+        # check if parameter files exist
+        fpath_params_individual = os.path.join(dir_params, PARAMS_INDIVIDUAL)
+        fpath_params_model = os.path.join(dir_params, PARAMS_MODEL)
+        assert os.path.isfile(fpath_params_individual), \
+            "Parameter file not found: %s." % PARAMS_INDIVIDUAL
+        assert os.path.isfile(fpath_params_model), \
+            "Parameter file not found: %s" % PARAMS_MODEL
```
Use an exception to indicate an error in the user input (here: the parameters file), and `assert` only to validate the code itself.

Here, this seems redundant because the functions that read the files later should raise the respective exception.
```suggestion
```

carschno commented on comma/__init__.py:
```diff
@@ -0,0 +1,64 @@
+    assert os.path.isfile(fpath), 'File not found: %s.' % fpath
+    with open(fpath) as file:
+        data = json.load(file)
+    return data
```
This function is redundant, it just calls `json.load()`. Whenever this function is called, `json.load()` should be called directly.

carschno commented on comma/model.py:
```diff
@@ -0,0 +1,190 @@
+            "Parameter file not found: %s" % PARAMS_MODEL
+
+        # check if all required model parameters are given 
+        params_model = read_json_as_dict(fpath_params_model)
```
```suggestion
        with open(fpath_params_model) as f:
             params_model = json.load(f)
```

carschno commented on comma/model.py:
```diff
@@ -0,0 +1,190 @@
+        # check if all required model parameters are given 
+        params_model = read_json_as_dict(fpath_params_model)
+        missing_model_params = set(self._required_params) - set(params_model.keys())
+        assert not missing_model_params, \
+            "Model parameter(s) not found: %s." % ", ".join(missing_model_params)
```
```suggestion
        if missing_model_params:
            raise ValueError("Model parameter(s) not found: %s." % ", ".join(missing_model_params))
```

carschno commented on comma/model.py:
```diff
@@ -0,0 +1,190 @@
+        assert not missing_model_params, \
+            "Model parameter(s) not found: %s." % ", ".join(missing_model_params)
+
+        # check if there is any overlap between lockdown intervals
+        lockdown = params_model["lockdown"]
+        lockdown_intervals = [l["interval"] for l in lockdown]
+        lockdown_intervals = sorted(lockdown_intervals, key=lambda li: li[0])
+        overlaps = [lockdown_intervals[i][1] >= lockdown_intervals[i + 1][0] \
+                    for i in range(len(lockdown_intervals) - 1)]
+        assert not any(overlaps), "Lockdown intervals have overlap: \n%s" % "\n".join(
+            ["%s - %s" % (str(lockdown_intervals[i]), str(lockdown_intervals[i + 1])) \
+             for i in range(len(overlaps)) if overlaps[i]]
```
I'm sure this can be more readable, but I am not clear about the exact functionality. I would suggest implementing a unit test first that checks the output for various inputs. Subsequently, the function can be optimized while having the test to ensure its functionality.

carschno commented on comma/model.py:
```diff
@@ -0,0 +1,190 @@
+        self.current_step: int = 0 # keep track of the current simulation step
+        self.lockdown_status: dict = {}
+
+    def setup(self, size: int, dir_params: str, use_ipf: bool = False):
```
The `__init__()` method is usually the place to set up the object; is there a reason why this is done in the separate `setup()` method here?

carschno commented on comma/__init__.py:
```diff
@@ -0,0 +1,64 @@
+PARAMS_MODEL = 'params_model.json'
+PARAMS_IPF_WEIGHTS = "ipf_weights.csv"
+
+
```
I tend to keep the `__init__.py` file mostly empty; I think the functions here can be either deleted or moved to a more descriptive location.

Here, it might even make sense to add a `Hypothesis` class that implements all the functions for reading and validating those files.


> A commonly seen issue is to add too much code to __init__.py files. When the project complexity grows, there may be sub-packages and sub-sub-packages in a deep directory structure. In this case, importing a single item from a sub-sub-package will require executing all __init__.py files met while traversing the tree.

https://python-docs.readthedocs.io/en/latest/writing/structure.html#packages

n400peanuts replied:
that's a great idea -- I'll make a to-do

Commit:
added test for take_actions()

n400peanuts committed:
Merge pull request #67 from covid19ABM/63-add-test-for-individualtake_actions

added test for take_actions()

Commit:
removed parameters_example folder

Commit:
added the correct matrices and updated params individual json

Commit:
refactoring comma

Commit:
no model test

Commit:
corrected with pylint

Commit:
corrected script based on pylint

Commit:
corrected scripts with flake8

Commit:
moved read hypotheses into hypotheses class

Commit:
test on individual improved for the latest changes

Commit:
removed params model as these are now integrated into the package

Commit:
demo individual updated

Commit:
removed mentions to the model setup.

n400peanuts committed:
Merge pull request #69 from covid19ABM/68-implement-changes-as-suggested-by-carstens-pr

removed parameters_example folder

Commit:
improved test individual class

Commit:
improved test ipf sampling

Commit:
improved test validate csv

Commit:
added test for model

n400peanuts committed:
Merge pull request #71 from covid19ABM/test_suite

improve test suite

Commit:
removed old reference to dir_params

Commit:
added sonarcloud

Commit:
added badge for sonar cloud

Commit:
added new test for the model module

Commit:
specified path for tests in sonarcloud

Commit:
specified package source in sonarcloud

Commit:
following walter advice

Commit:
second advice on sonarcloud

Commit:
added pytest-cov

Astrid-p committed:
Add files via upload

Hi Eva,

Thank you for sparing your time for me today! The attached folder includes all the required matrices, including the action-effect and the corrected versions of action-probability matrices. Please dump the previously uploaded files and use these instead. I am sorry for the inconvenience 😥.

best wishes, 
AP

Commit:
CITATION file added

Commit:
added license badge

Commit:
added contributions guidelines

Commit:
code of conduct

Commit:
updated readme badge

Commit:
updated rst readme

Commit:
updated readme

Commit:
fix images in docs

Commit:
fix images in docs

Commit:
fix images in documentation

Commit:
fix documentation

Astrid-p committed:
Delete parameters_complete_24Aug directory

error in float format

Astrid-p committed:
correcting float format

Astrid-p committed:
Delete parameters_complete_28Aug directory

Astrid-p committed:
corrected-format_2

Astrid-p committed:
Delete lockdown_absent.csv

Astrid-p committed:
Delete effects_on_mh_HardLD.csv

Astrid-p committed:
Delete effects_on_mh_withoutLD.csv

Astrid-p committed:
Delete lockdown_easy.csv

Astrid-p committed:
Delete lockdown_hard.csv

Astrid-p committed:
Delete lockdown_medium.csv

Astrid-p committed:
Delete effects_on_mh_MediumLD.csv

Astrid-p committed:
Delete effects_on_mh_EasyLD.csv

Astrid-p committed:
correcting-decimal-format-2

Astrid-p committed:
correcting-decimal-format-2

Astrid-p committed:
Update and rename effects_on_mh_EasyLD.csv to actions_effects_on_mh_easy.csv

Astrid-p committed:
Update and rename effects_on_mh_withoutLD.csv to actions_effects_on_mh_absent.csv

Astrid-p committed:
Update and rename effects_on_mh_HardLD.csv to actions_effects_on_mh_hard.csv

Astrid-p committed:
Update and rename effects_on_mh_MediumLD.csv to actions_effects_on_mh_medium.csv

Commit:
refactor code and correct matrices

Commit:
add parameters

Commit:
gergmerge necessarye branch 'main' into Complete-matrix-file

Commit:
tests updated

Commit:
remove simulation results

Commit:
individual class updated

Commit:
fixed type set in action_effects

Commit:
removed old notebook

Commit:
fixed duplicated code

Commit:
added coverage

Commit:
removed commented code

n400peanuts committed:
Merge pull request #72 from covid19ABM/Complete-matrix-file

Add files via upload

Commit:
type hints added

Commit:
update docs with changelog

Commit:
fix hyperlink

Commit:
fix hyperlinks

Commit:
diagram updated according to the latest package changes

Commit:
added download covid data functionality in hypothesis

Commit:
varied examples in the test

Commit:
new documentation about covid-19 positivity

Commit:
updated changelog

Commit:
positivity estimation implemented

Commit:
updated documentation

Commit:
refactor

Commit:
update code to reflect new numpy random generator

Commit:
adjusted tolerance to new numpy random generator

Commit:
make quality gate happy

Commit:
set unpredictable random seed

Commit:
updated seed

Commit:
remove code smells

Commit:
removed inplace=True

Commit:
changelog updated

Commit:
updated whatsnew

Commit:
removed unused notebook

Commit:
update diagram

Commit:
code restyle
;

Commit:
black style adjusted

Commit:
change integrated into changelog and readmes

n400peanuts committed:
Merge pull request #74 from covid19ABM/black

code restyle

carschno commented:
This has been processed.
